In [ ]:
import pandas as pd
import numpy as np
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)
pd.set_option('display.width', None)

# In this version I deleted those bad tag data (Those who receives a tag with number that is even greater than the total number of ratings)

In [ ]:
import pandas as pd 

# load datasets
rmp_num = pd.read_csv('rmpCapstoneNum.csv', header=None)
rmp_qual = pd.read_csv('rmpCapstoneQual.csv', header=None)
rmp_tags = pd.read_csv('rmpCapstoneTags.csv', header=None)

# assign column names
rmp_num.columns = [
    "Average Rating",
    "Average Difficulty",
    "Number of ratings",
    "Received a “pepper”?",
    "The proportion of students that said they would take the class again",
    "The number of ratings coming from online classes",
    "Male gender",
    "Female"
]

rmp_qual.columns = [
    "Major/Field",
    "University",
    "US State"
]

rmp_tags.columns = [
    "Tough grader",
    "Good feedback",
    "Respected",
    "Lots to read",
    "Participation matters",
    "Don’t skip class or you will not pass",
    "Lots of homework",
    "Inspirational",
    "Pop quizzes!",
    "Accessible",
    "So many papers",
    "Clear grading",
    "Hilarious",
    "Test heavy",
    "Graded by few things",
    "Amazing lectures",
    "Caring",
    "Extra credit",
    "Group projects",
    "Lecture heavy"
]

# merge and clean data
rmp = pd.concat([rmp_num, rmp_qual, rmp_tags], axis=1)
print(f"original length: {len(rmp)}")

rmp = rmp.dropna(subset=["Average Rating"]).drop_duplicates()
print(f"after dropping nan in Average Rating : {len(rmp)}")

tag_columns = rmp_tags.columns
rmp = rmp[rmp.loc[:, "Number of ratings"] >= rmp.loc[:, rmp_tags.columns].max(axis=1)]
print(f"then dropped bad tag data: {len(rmp)}")

# Bayesian adjustments for average rating and difficulty
# calculate prior means
prior_mean_rating = rmp["Average Rating"].mean()
prior_mean_difficulty = rmp["Average Difficulty"].mean()

# use median of 'Number of ratings' as strength of prior
strength_of_prior = rmp["Number of ratings"].median()

# apply Bayesian adjustment for ‘Average Rating’
rmp["Average Rating (Adjusted)"] = (
    (strength_of_prior * prior_mean_rating + 
     rmp["Number of ratings"] * rmp["Average Rating"]) /
    (strength_of_prior + rmp["Number of ratings"])
)

# apply Bayesian adjustment for ‘Average Difficulty’
rmp["Average Difficulty (Adjusted)"] = (
    (strength_of_prior * prior_mean_difficulty + 
     rmp["Number of ratings"] * rmp["Average Difficulty"]) /
    (strength_of_prior + rmp["Number of ratings"])
)

# normalize tag columns by 'Number of ratings'
tag_columns = rmp_tags.columns
for tag in tag_columns:
    rmp[f"{tag} (Normalized)"] = rmp[tag] / rmp["Number of ratings"]

# calculate total number of tags
rmp["Total Number of Tags"] = rmp[tag_columns].sum(axis=1)

# calculate average number of tags per rating
rmp["Average Tags per Rating"] = rmp["Total Number of Tags"] / rmp["Number of ratings"]

# reorder columns
column_order = [
    # numerical columns
    "Average Rating",
    "Average Rating (Adjusted)",
    "Average Difficulty",
    "Average Difficulty (Adjusted)",
    "Number of ratings",
    "Received a “pepper”?",
    "The proportion of students that said they would take the class again",
    "The number of ratings coming from online classes",
    "Male gender",
    "Female",
    
    # qualitative columns
    "Major/Field",
    "University",
    "US State",

    # tag columns
    *tag_columns,  # original tag columns
    *[f"{tag} (Normalized)" for tag in tag_columns],  # normalized tag columns
    "Total Number of Tags",
    "Average Tags per Rating"
]

# apply the column order
rmp = rmp[column_order]

# save the cleaned and adjusted data
rmp.to_csv(f'rmpCapstoneAdjusted_{len(rmp)}.csv', index=False)

original length: 89893
after dropping nan in Average Rating : 69989
then dropped bad tag data: 69706


# The following is the 69989 version of data cleaning

In [ ]:
import pandas as pd

# load datasets
rmp_num = pd.read_csv('rmpCapstoneNum.csv', header=None)
rmp_qual = pd.read_csv('rmpCapstoneQual.csv', header=None)
rmp_tags = pd.read_csv('rmpCapstoneTags.csv', header=None)

# assign column names
rmp_num.columns = [
    "Average Rating",
    "Average Difficulty",
    "Number of ratings",
    "Received a “pepper”?",
    "The proportion of students that said they would take the class again",
    "The number of ratings coming from online classes",
    "Male gender",
    "Female"
]

rmp_qual.columns = [
    "Major/Field",
    "University",
    "US State"
]

rmp_tags.columns = [
    "Tough grader",
    "Good feedback",
    "Respected",
    "Lots to read",
    "Participation matters",
    "Don’t skip class or you will not pass",
    "Lots of homework",
    "Inspirational",
    "Pop quizzes!",
    "Accessible",
    "So many papers",
    "Clear grading",
    "Hilarious",
    "Test heavy",
    "Graded by few things",
    "Amazing lectures",
    "Caring",
    "Extra credit",
    "Group projects",
    "Lecture heavy"
]

# merge and clean data
rmp = pd.concat([rmp_num, rmp_qual, rmp_tags], axis=1)
rmp = rmp.dropna(subset=["Average Rating"]).drop_duplicates()

# Bayesian adjustments for average rating and difficulty
# calculate prior means
prior_mean_rating = rmp["Average Rating"].mean()
prior_mean_difficulty = rmp["Average Difficulty"].mean()

# use median of 'Number of ratings' as strength of prior
strength_of_prior = rmp["Number of ratings"].median()

# apply Bayesian adjustment for ‘Average Rating’
rmp["Average Rating (Adjusted)"] = (
    (strength_of_prior * prior_mean_rating + 
     rmp["Number of ratings"] * rmp["Average Rating"]) /
    (strength_of_prior + rmp["Number of ratings"])
)

# apply Bayesian adjustment for ‘Average Difficulty’
rmp["Average Difficulty (Adjusted)"] = (
    (strength_of_prior * prior_mean_difficulty + 
     rmp["Number of ratings"] * rmp["Average Difficulty"]) /
    (strength_of_prior + rmp["Number of ratings"])
)

# normalize tag columns by 'Number of ratings'
tag_columns = rmp_tags.columns
for tag in tag_columns:
    rmp[f"{tag} (Normalized)"] = rmp[tag] / rmp["Number of ratings"]

# calculate total number of tags
rmp["Total Number of Tags"] = rmp[tag_columns].sum(axis=1)

# calculate average number of tags per rating
rmp["Average Tags per Rating"] = rmp["Total Number of Tags"] / rmp["Number of ratings"]

# reorder columns
column_order = [
    # numerical columns
    "Average Rating",
    "Average Rating (Adjusted)",
    "Average Difficulty",
    "Average Difficulty (Adjusted)",
    "Number of ratings",
    "Received a “pepper”?",
    "The proportion of students that said they would take the class again",
    "The number of ratings coming from online classes",
    "Male gender",
    "Female",
    
    # qualitative columns
    "Major/Field",
    "University",
    "US State",

    # tag columns
    *tag_columns,  # original tag columns
    *[f"{tag} (Normalized)" for tag in tag_columns],  # normalized tag columns
    "Total Number of Tags",
    "Average Tags per Rating"
]

# apply the column order
rmp = rmp[column_order]

# save the cleaned and adjusted data
rmp.to_csv(f'rmpCapstoneAdjusted_{len(rmp)}.csv', index=False)

In [ ]:
rmp["Number of ratings"].value_counts()

1.0      17633
2.0      11843
3.0       8577
4.0       6583
5.0       4791
6.0       3736
7.0       2846
8.0       2267
9.0       1887
10.0      1495
11.0      1144
12.0       966
13.0       831
14.0       726
15.0       534
16.0       523
17.0       402
18.0       374
19.0       298
20.0       281
21.0       228
22.0       210
23.0       194
24.0       134
25.0       126
27.0       113
26.0       110
28.0        81
29.0        79
30.0        76
31.0        71
32.0        70
33.0        59
35.0        52
36.0        49
38.0        41
37.0        36
34.0        35
39.0        31
40.0        27
41.0        26
42.0        25
43.0        23
47.0        20
45.0        18
44.0        16
52.0        15
48.0        14
51.0        13
63.0        12
46.0        12
68.0        12
53.0        11
56.0        10
54.0        10
49.0        10
50.0         9
67.0         8
70.0         7
61.0         7
58.0         7
60.0         7
65.0         6
55.0         5
57.0         5
93.0         5
77.0      